### DISCUSSION TITLE
InputDevice pull up parameter not working gpiozero raspberry pi 5

### DISCUSSION THREAD

**Discussion by sayedMbo** [gpiozero/gpiozero#1155] [Q&A] [answered]
### InputDevice pull up parameter not working gpiozero raspberry pi

Pi model:  Pi 5 
to test this I made a simple circuit with a Led,  **Now the Led is turning on** but when I try to print the pin's value it's still 0 also the is_active method also gives False

```
from gpiozero import InputDevice

# Initialize the InputDevice on pin 17 with pull_up set to True
pin = InputDevice(17, pull_up=True)

# Function to check if the pin is turned on
def check_pin_status():
    if pin.value == 1:
        print("The pin is turned on (value is 1).")
    else:
        print("The pin is turned off (value is 0).") # this gives 0

   print(pin.is_active) # this also gives False 

# Check the pin status
check_pin_status()

```

**steps to make the circuit**
Connect a wire to GPIO pin 17 on the Raspberry Pi.
Connect the other end of that wire to one end of a resistor (e.g., 220 ohms).
Connect the other end of the resistor to the anode (longer leg) of an LED.
Connect the cathode (shorter leg) of the LED to a ground pin on the Raspberry Pi using another wire.

**What my goal is where is it going Wrong**
what I am trying to do with InputDevice  is that the pin 17 connected wire is connected to a button, and the other side of the button is grounded. When  pull_up=True I read the pin without pressing the button it should read 1 and when the button is pressed it should read 0 . BUT both pin.value  is 0  and also pin.is_active = False all the time whether the button is pushed or not

**Comment by lurch** [comment]
"Connect the cathode (shorter leg) of the LED to a ground pin on the Raspberry Pi"

"the other side of the button is grounded"

Huh?!? In one part of your issue you talk about using an LED, and in the other part of your issue you talk about using a button? (Also, it doesn't make any sense to try using an LED with `InputDevice`.)

But the good news is that with GpioZero you don't need to mess about with the low-level [`InputDevice`](https://gpiozero.readthedocs.io/en/latest/api_input.html#gpiozero.InputDevice), as GpioZero has higher-level [`LED`](https://gpiozero.readthedocs.io/en/latest/api_output.html#gpiozero.LED) and [`Button`](https://gpiozero.readthedocs.io/en/latest/api_input.html#gpiozero.Button) classes which are much more convenient to use. See https://gpiozero.readthedocs.io/en/latest/recipes.html#button for a nicely-documented example.

**Comment by sayedMbo** [comment]
Thank you @lurch  for your reply and sorry for any confusion.

I have a 17 pin connected, and I have set it up as follows:


`pin = InputDevice(17, pull_up=True)
`
This configuration should pull the pin up to 3.3V internally. Therefore:

- pin.value should return 1, but it is returning 0.
- pin.is_active should be True, but it is False.

Can you help me understand why this is happening?

**Reply by lurch** [reply]
Do you **actually** have a button connected to pin 17 now, or are you still trying to "test" it with an LED?
Why not just use the Button class as I mentioned in my last comment?

**Reply by sayedMbo** [reply]
I tested the setup with an LED to verify if pull_up=True is functioning correctly since I couldn't confirm it using the pin.value and pin.is_active methods. The LED worked as expected.

However, my main goal is to work with a button. Here's what I'm trying to achieve:

- Enable the pull-up resistor on the pin by setting pull_up=True.
- Confirm that the pin's value is initially 1.
- When the button, which is grounded on one end, is pressed, it will change the pin's value to 0, acting as a switch.
-  Its like a switch or pull up resistor thing. 

I want to avoid using the  pin number 1 that defaults to 3V or a value of 1 without setting pull_up=True.

**Reply by lurch** [reply, marked as answer]
> However, my main goal is to work with a button. Here's what I'm trying to achieve:

Yes, that's exactly how the `Button` class in GpioZero works. See the wiring diagram in the example button recipe I linked to earlier.

> I want to avoid using the pin number 1 that defaults to 3V or a value of 1 without setting pull_up=True.

Yes, if you look at the Button [API docs](https://gpiozero.readthedocs.io/en/latest/api_input.html#gpiozero.Button) you can see that it says that the `pull_up` parameter defaults to `True`, i.e.
```python
button = Button(17)
```
does exactly the same thing as
```python
button = Button(17, pull_up=True)
```

As Ben says, in GpioZero the Button's [`value`](https://gpiozero.readthedocs.io/en/latest/api_input.html#gpiozero.Button.value) returns True if the button is being pressed, regardless of whether it's operating in "pull-up mode" or "pull-down mode". If you **really** want to see the state of the underlying [pin](https://gpiozero.readthedocs.io/en/latest/api_pins.html#gpiozero.Pin), instead of looking at `button.value` you could look at `button.pin.state`.

**Comment by bennuttall** [comment]
You are mixing up the device's value with the pin state.

GPIO pins can be high and low but gpiozero abstracts the pin state from the device. The value of a button is 1 if the button is pressed or 0 if it is not pressed, regardless of the pull-up state.

**Comment by sayedMbo** [comment]
Thanks, @lurch  and @bennuttall   for your replies. I understand what you are saying.

I was trying to implement Rpi.GPIO in gpiozero, which led me to use:
`pin = InputDevice(17, pull_up=True)
`
I was wondering why its value isn't coming out to be 1 when pull_up=True, but I understand your input and maybe will try that.

<img width="458" alt="w" src="https://github.com/gpiozero/gpiozero/assets/142673964/de9401ea-a6e6-43cf-9f72-01fb3a95e92f">

**Reply by bennuttall** [reply]
It really doesn't make sense to implement RPi.GPIO in gpiozero. gpiozero is a layer of abstraction on top of a number of pin libraries. If you want to use the RPi.GPIO syntax (I'm not sure why), use something like this: https://github.com/waveform80/rpi-lgpio
